File: crates/eg-compute/src/solve/verify/status.rs
//! Does the claimed status follow from the verified evidence?

use super::error::{StatusDefect, Verdict, VerifyError};
use super::tree::Evidence;
use crate::solve::certificate::{Certificate, SolveStatus};
use crate::solve::model::RowId;
use crate::solve::scalar::Scalar;

/// Facts established independently of the claimed status.
pub(super) struct Verified {
    /// The incumbent's recomputed scalar objective, when it is feasible.
    pub(super) incumbent: Option<i128>,
    pub(super) evidence: Evidence,
    /// Whether the proof was a full tree (rather than a root dual).
    pub(super) tree: bool,
}

fn require(condition: bool, defect: StatusDefect) -> Result<(), VerifyError> {
    if condition {
        Ok(())
    } else {
        Err(VerifyError::Status(defect))
    }
}

pub(super) fn judge(
    certificate: &Certificate,
    verified: &Verified,
) -> Result<Verdict, VerifyError> {
    let claimed = certificate.lower_bound.map(Scalar::get);
    match &certificate.status {
        SolveStatus::Optimal => optimal(claimed, verified),
        SolveStatus::OptimalByDeterministicSearch { nodes_expanded } => {
            require(
                *nodes_expanded == certificate.nodes_expanded,
                StatusDefect::NodeCountMismatch,
            )?;
            optimal_by_search(claimed, verified)
        }
        SolveStatus::FeasibleWithGap { gap } => with_gap(certificate, gap.get(), verified),
        SolveStatus::Infeasible { core } => infeasible(claimed, core, verified),
        SolveStatus::InfeasibleByDeterministicSearch { nodes_expanded } => {
            require(
                *nodes_expanded == certificate.nodes_expanded,
                StatusDefect::NodeCountMismatch,
            )?;
            infeasible_by_search(claimed, verified)
        }
        SolveStatus::BudgetExhausted => exhausted(certificate, claimed, verified),
    }
}

fn incumbent(verified: &Verified) -> Result<i128, VerifyError> {
    verified
        .incumbent
        .ok_or(VerifyError::Status(StatusDefect::MissingIncumbent))
}

/// A claimed lower bound that the verified leaves support.
fn supported_bound(claimed: Option<i128>, verified: &Verified) -> Result<i128, VerifyError> {
    let bound = claimed.ok_or(VerifyError::Status(StatusDefect::LowerBoundMismatch))?;
    let proven = verified
        .evidence
        .min_bound
        .ok_or(VerifyError::Status(StatusDefect::NoBoundLeaf))?;
    require(bound <= proven, StatusDefect::LowerBoundAboveProof)?;
    Ok(bound)
}

fn optimal(claimed: Option<i128>, verified: &Verified) -> Result<Verdict, VerifyError> {
    let best = incumbent(verified)?;
    let proven = verified
        .evidence
        .min_bound
        .ok_or(VerifyError::Status(StatusDefect::NoBoundLeaf))?;
    require(proven >= best, StatusDefect::BoundBelowIncumbent)?;
    require(claimed == Some(best), StatusDefect::LowerBoundMismatch)?;
    Ok(Verdict::ProvenOptimal {
        objective: Scalar::new(best),
    })
}

fn optimal_by_search(claimed: Option<i128>, verified: &Verified) -> Result<Verdict, VerifyError> {
    require(!verified.tree, StatusDefect::WrongProofKind)?;
    let best = incumbent(verified)?;
    let bound = supported_bound(claimed, verified)?;
    require(bound <= best, StatusDefect::LowerBoundMismatch)?;
    Ok(Verdict::OptimalityRequiresReplay {
        objective: Scalar::new(best),
        lower_bound: Scalar::new(bound),
    })
}

fn with_gap(
    certificate: &Certificate,
    gap: i128,
    verified: &Verified,
) -> Result<Verdict, VerifyError> {
    let budget = certificate.config.node_budget();
    require(
        certificate.nodes_expanded == budget,
        StatusDefect::NodeCountMismatch,
    )?;
    let best = incumbent(verified)?;
    let bound = supported_bound(certificate.lower_bound.map(Scalar::get), verified)?;
    require(
        gap > 0 && best.checked_sub(bound) == Some(gap),
        StatusDefect::GapMismatch,
    )?;
    require(
        gap <= certificate.config.accepted_gap(),
        StatusDefect::GapNotAccepted,
    )?;
    Ok(Verdict::ProvenGap {
        incumbent: Scalar::new(best),
        lower_bound: Scalar::new(bound),
    })
}

fn no_incumbent_no_bound(claimed: Option<i128>, verified: &Verified) -> Result<(), VerifyError> {
    require(
        verified.incumbent.is_none(),
        StatusDefect::UnexpectedIncumbent,
    )?;
    require(claimed.is_none(), StatusDefect::LowerBoundMismatch)
}

fn infeasible(
    claimed: Option<i128>,
    core: &[RowId],
    verified: &Verified,
) -> Result<Verdict, VerifyError> {
    no_incumbent_no_bound(claimed, verified)?;
    require(verified.tree, StatusDefect::WrongProofKind)?;
    require(
        verified.evidence.bound_leaves == 0,
        StatusDefect::WrongProofKind,
    )?;
    require(
        core == verified.evidence.infeasibility_rows,
        StatusDefect::CoreMismatch,
    )?;
    Ok(Verdict::ProvenInfeasible {
        core: core.to_vec(),
    })
}

fn infeasible_by_search(
    claimed: Option<i128>,
    verified: &Verified,
) -> Result<Verdict, VerifyError> {
    no_incumbent_no_bound(claimed, verified)?;
    require(!verified.tree, StatusDefect::WrongProofKind)?;
    Ok(Verdict::InfeasibilityRequiresReplay)
}

fn exhausted(
    certificate: &Certificate,
    claimed: Option<i128>,
    verified: &Verified,
) -> Result<Verdict, VerifyError> {
    let budget = certificate.config.node_budget();
    require(
        certificate.nodes_expanded == budget,
        StatusDefect::NodeCountMismatch,
    )?;
    let bound = match claimed {
        Some(_) => Some(supported_bound(claimed, verified)?),
        None => None,
    };
    if let (Some(best), Some(bound)) = (verified.incumbent, bound) {
        let gap = best.checked_sub(bound).unwrap_or(i128::MAX);
        require(
            gap > certificate.config.accepted_gap(),
            StatusDefect::GapAccepted,
        )?;
    }
    Ok(Verdict::Unresolved {
        incumbent: verified.incumbent.map(Scalar::new),
        lower_bound: bound.map(Scalar::new),
    })
}

File: epistemic_graph/decision/_status.py
"""Judge a certificate's claimed status against independently verified facts."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from ._certificate import CertificateError, Model
from ._proof import Evidence, incumbent_value, root, walk


@dataclass(frozen=True)
class Verdict:
    """What a certificate was verified to establish.

    ``kind`` is one of ``proven_optimal``, ``optimality_requires_replay``,
    ``proven_gap``, ``proven_infeasible``, ``infeasibility_requires_replay`` or
    ``unresolved``, exactly as the engine's verifier names them.
    """

    kind: str
    objective: int | None = None
    lower_bound: int | None = None
    core: tuple[int, ...] = ()

    @property
    def supports_an_answer(self) -> bool:
        """Whether this verdict supports a ``Solved`` assembly."""
        return self.kind in {
            "proven_optimal",
            "optimality_requires_replay",
            "proven_gap",
        }


@dataclass(frozen=True)
class _Facts:
    incumbent: int | None
    evidence: Evidence
    tree: bool
    claimed: int | None
    certificate: dict[str, Any]


def _require(condition: bool, defect: str) -> None:
    if not condition:
        raise CertificateError(f"status defect: {defect}")


def _value(value: int | None, defect: str) -> int:
    if value is None:
        raise CertificateError(f"status defect: {defect}")
    return value


def _incumbent(facts: _Facts) -> int:
    return _value(facts.incumbent, "missing incumbent")


def _supported_bound(facts: _Facts) -> int:
    claimed = _value(facts.claimed, "lower bound mismatch")
    proven = _value(facts.evidence.min_bound, "no bound leaf")
    _require(claimed <= proven, "lower bound above proof")
    return claimed


def _nodes(facts: _Facts, body: dict[str, Any]) -> None:
    _require(
        body["nodes_expanded"] == facts.certificate["nodes_expanded"],
        "node count mismatch",
    )


def _budget_spent(facts: _Facts) -> None:
    spent = (
        facts.certificate["nodes_expanded"]
        == facts.certificate["config"]["node_budget"]
    )
    _require(spent, "node count mismatch")


def _optimal(facts: _Facts, _: Any) -> Verdict:
    best = _incumbent(facts)
    proven = _value(facts.evidence.min_bound, "no bound leaf")
    _require(proven >= best, "bound below incumbent")
    _require(facts.claimed == best, "lower bound mismatch")
    return Verdict("proven_optimal", objective=best)


def _optimal_by_search(facts: _Facts, body: Any) -> Verdict:
    _nodes(facts, body)
    _require(not facts.tree, "wrong proof kind")
    best = _incumbent(facts)
    bound = _supported_bound(facts)
    _require(bound <= best, "lower bound mismatch")
    return Verdict("optimality_requires_replay", objective=best, lower_bound=bound)


def _with_gap(facts: _Facts, body: Any) -> Verdict:
    _budget_spent(facts)
    best = _incumbent(facts)
    bound = _supported_bound(facts)
    gap = int(body["gap"])
    _require(gap > 0 and best - bound == gap, "gap mismatch")
    _require(
        gap <= int(facts.certificate["config"]["accepted_gap"]), "gap not accepted"
    )
    return Verdict("proven_gap", objective=best, lower_bound=bound)


def _no_incumbent_no_bound(facts: _Facts) -> None:
    _require(facts.incumbent is None, "unexpected incumbent")
    _require(facts.claimed is None, "lower bound mismatch")


def _infeasible(facts: _Facts, body: Any) -> Verdict:
    _no_incumbent_no_bound(facts)
    _require(facts.tree and facts.evidence.bound_leaves == 0, "wrong proof kind")
    _require(list(body["core"]) == facts.evidence.infeasibility_rows, "core mismatch")
    return Verdict("proven_infeasible", core=tuple(body["core"]))


def _infeasible_by_search(facts: _Facts, body: Any) -> Verdict:
    _nodes(facts, body)
    _no_incumbent_no_bound(facts)
    _require(not facts.tree, "wrong proof kind")
    return Verdict("infeasibility_requires_replay")


def _exhausted(facts: _Facts, _: Any) -> Verdict:
    _budget_spent(facts)
    bound = None if facts.claimed is None else _supported_bound(facts)
    if facts.incumbent is not None and bound is not None:
        accepted = int(facts.certificate["config"]["accepted_gap"])
        _require(facts.incumbent - bound > accepted, "gap accepted")
    return Verdict("unresolved", objective=facts.incumbent, lower_bound=bound)


_JUDGES = {
    "optimal": _optimal,
    "optimal_by_deterministic_search": _optimal_by_search,
    "feasible_with_gap": _with_gap,
    "infeasible": _infeasible,
    "infeasible_by_deterministic_search": _infeasible_by_search,
    "budget_exhausted": _exhausted,
}


def _status(certificate: dict[str, Any]) -> tuple[str, Any]:
    status = certificate["status"]
    if isinstance(status, str):
        return status, None
    ((kind, body),) = status.items()
    return kind, body


def verify_certificate(
    model_spec: dict[str, Any], certificate: dict[str, Any]
) -> Verdict:
    """Verify ``certificate`` against ``model_spec`` without trusting the solver.

    Raises :class:`CertificateError` naming the first defect found.
    """
    model = Model(model_spec)
    if certificate["model_digest"] != model.digest:
        raise CertificateError("the certificate was issued for a different model")
    if certificate["nodes_expanded"] > certificate["config"]["node_budget"]:
        raise CertificateError("status defect: node count mismatch")
    incumbent = incumbent_value(model, certificate)
    ((proof_kind, proof),) = certificate["proof"].items()
    tree = proof_kind == "tree"
    evidence = walk(model, proof["nodes"]) if tree else root(model, proof["dual"])
    claimed = certificate.get("lower_bound")
    facts = _Facts(
        incumbent=incumbent,
        evidence=evidence,
        tree=tree,
        claimed=None if claimed is None else int(claimed),
        certificate=certificate,
    )
    kind, body = _status(certificate)
    return _JUDGES[kind](facts, body)
